File: src/main/java/com/github/tammo/fun/parse/AntlrParser.java
package com.github.tammo.fun.parse;

import com.github.tammo.FunLexer;
import com.github.tammo.FunParser;
import com.github.tammo.fun.ast.SyntaxNode.CompilationUnit;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;

public class AntlrParser implements Parser {

    private final ASTBuilder astBuilder = new ASTBuilder();

    @Override
    public CompilationUnit parse(String input) {
        final CharStream charStream = CharStreams.fromString(input);

        final FunLexer lexer = new FunLexer(charStream);
        final CommonTokenStream tokens = new CommonTokenStream(lexer);

        final FunParser funParser = new FunParser(tokens);
        return (CompilationUnit) astBuilder.visitCompilationUnit(funParser.compilationUnit());
    }

}

File: src/main/scala/com/github/tammo/frontend/parse/AntlrParser.scala
package com.github.tammo.frontend.parse

import com.github.tammo.diagnostics.SourceFile
import com.github.tammo.frontend.ast.SyntaxTree
import com.github.tammo.frontend.ast.SyntaxTree.CompilationUnit
import com.github.tammo.{FunLexer, FunParser}
import org.antlr.v4.runtime.{CharStreams, CommonTokenStream}

object AntlrParser extends Parser {

  override def parse(sourceFile: SourceFile): CompilationUnit = {
    val charStream = CharStreams.fromString(sourceFile.content)

    val lexer = FunLexer(charStream)
    val tokens = CommonTokenStream(lexer)

    val funParser = new FunParser(tokens)

    val syntaxTreeBuilder = SyntaxTreeBuilder(sourceFile)

    syntaxTreeBuilder
      .visitCompilationUnit(funParser.compilationUnit())
      .asInstanceOf[CompilationUnit]
  }

}
